sql: cursors which buffer notices panic

**Describe the problem**

When we try buffer a notice for a cursor, we panic. This should probably no-op instead (or forward the notice through).

**To Reproduce**

Run
```
BEGIN; declare a cursor for select * from crdb_internal.gossip_network;
```

You'll see:
```

*
* ERROR: a panic has occurred!
* panic: unimplemented
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql.(*streamingCommandResult).BufferNotice
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_io.go:943
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).BufferClientNotice
*   | 	github.com/cockroachdb/cockroach/pkg/sql/notice.go:55
*   | github.com/cockroachdb/cockroach/pkg/sql.glob..func72
*   | 	github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:4410
*   | github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).getPlanInfo.func1.1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:638
*   | github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:127
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_arm64.s:1172
* Wraps: (2) panic: unimplemented
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
panic: unimplemented [recovered]
	panic: unimplemented

```

CockroachDB then panics and dies.

Jira issue: CRDB-22841

--- a/pkg/sql/conn_io.go
+++ b/pkg/sql/conn_io.go
@@ -935,12 +935,12 @@ func (r *streamingCommandResult) SetColumns(ctx context.Context, cols colinfo.Re
 
 // BufferParamStatusUpdate is part of the RestrictedCommandResult interface.
 func (r *streamingCommandResult) BufferParamStatusUpdate(key string, val string) {
-	panic("unimplemented")
+	// Unimplemented: the internal executor does not support status updated.
 }
 
 // BufferNotice is part of the RestrictedCommandResult interface.
 func (r *streamingCommandResult) BufferNotice(notice pgnotice.Notice) {
-	panic("unimplemented")
+	// Unimplemented: the internal executor does not support notices.
 }
 
 // ResetStmtType is part of the RestrictedCommandResult interface.

--- a/pkg/sql/logictest/testdata/logic_test/cursor
+++ b/pkg/sql/logictest/testdata/logic_test/cursor
@@ -643,3 +643,11 @@ DECLARE foo CURSOR WITH HOLD FOR SELECT 1
 # ROLLBACK is fine, since it renders the cursor unusable anyway.
 statement ok
 ROLLBACK
+
+# Regression test for using a SQL cursor that buffers a notice.
+# See https://github.com/cockroachdb/cockroach/issues/94344
+statement ok
+BEGIN;
+declare a cursor for select * from crdb_internal.gossip_network;
+FETCH 1 FROM a;
+COMMIT

